kotlin-coroutines-end/app/src/main/java/com/example/android/kotlincoroutines/main/RefreshMainDataWork.kt

```diff
@@ -18,44 +18,40 @@ package com.example.android.kotlincoroutines.main
 
 import android.content.Context
 import androidx.annotation.WorkerThread
+import androidx.work.CoroutineWorker
 import androidx.work.Worker
 import androidx.work.WorkerParameters
-import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.*
 
 /**
- * Worker job to refresh refresh titles from the network while the app is in the background.
+ * Worker job to refresh titles from the network while the app is in the background.
  *
  * WorkManager is a library used to enqueue work that is guaranteed to execute after its constraints
  * are met. It can run work even when the app is in the background, or not running.
  */
-class RefreshMainDataWork(context: Context, params: WorkerParameters) : Worker(context, params) {
+class RefreshMainDataWork(context: Context, params: WorkerParameters) :
+        CoroutineWorker(context, params) {
+
+    // CoroutineWorker defaults to Dispatchers.Default
+    // We want to execute this on Dispatchers.IO
+    override val coroutineContext = Dispatchers.IO
 
     /**
-     * Do our actual processing for the worker.
+     * Refresh the title from the network using [TitleRepository]
      *
      * WorkManager will call this method from a background thread. It may be called even
      * after our app has been terminated by the operating system, in which case [WorkManager] will
      * start just enough to run this [Worker].
      */
-    override fun doWork(): Result {
-        return refreshTitle()
-    }
-
-    /**
-     * Refresh the title from the network using [TitleRepository]
-     */
-    @WorkerThread
-    fun refreshTitle(): Result {
-        return runBlocking {
-            val database = getDatabase(applicationContext)
-            val repository = TitleRepository(MainNetworkImpl, database.titleDao)
+    override suspend fun doWork(): Result = coroutineScope {
+        val database = getDatabase(applicationContext)
+        val repository = TitleRepository(MainNetworkImpl, database.titleDao)
 
-            try {
-                repository.refreshTitle()
-                Result.success()
-            } catch (error: TitleRefreshError) {
-                Result.failure()
-            }
+        try {
+            repository.refreshTitle()
+            Result.success()
```

Comment:
These can be `return` without the `coroutineScope` lambda.

Reply:
Please check the CoroutineWorker implementation to be sure that this make sense:
https://github.com/aosp-mirror/platform_frameworks_support/blob/4d7787e10d8a8637250f4e8f2b43ae1b3ccb2555/work/workmanager-ktx/src/main/java/androidx/work/CoroutineWorker.kt#L64

Reply:
Aye - since doWork is a regular suspend function you can map it to a regular function. No need for scopes, launch, or async here!

Reply:
I've removed the imports and updated the PR.